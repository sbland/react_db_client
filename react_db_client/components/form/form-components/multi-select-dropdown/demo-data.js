export const demoOptions = [
    {
      uid: 'a',
      label: 'A',
    },
    {
      uid: 'b',
      label: 'B',
    },
    {
      uid: 'c',
      label: 'C',
    },
    {
      uid: 'd',
      label: 'D',
    },
    {
      uid: 'e',
      label: 'E',
    },
  ];
