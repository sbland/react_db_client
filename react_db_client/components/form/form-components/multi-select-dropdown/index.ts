export { MultiSelectDropdown } from './multi-select-dropdown';
