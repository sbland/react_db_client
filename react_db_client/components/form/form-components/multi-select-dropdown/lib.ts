export interface IOpt {
  uid: string | number;
  label: string;
}