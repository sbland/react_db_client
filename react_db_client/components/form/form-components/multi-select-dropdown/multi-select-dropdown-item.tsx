import React from 'react';
import PropTypes from 'prop-types';
import { IOpt } from './lib';

export interface IMultiSelectDropdownItemProps extends IOpt {
  handleSelect: (uid: string | number) => void;
  isSelected: boolean;
}

export const MultiSelectDropdownItem = ({
  uid,
  label,
  handleSelect,
  isSelected,
}: IMultiSelectDropdownItemProps) => (
  <li className="multiSelectDropdown_item">
    <button
      type="button"
      className={isSelected ? 'button-two selected' : 'button-one notSelected'}
      onClick={() => handleSelect(uid)}
    >
      {label}
    </button>
  </li>
);

MultiSelectDropdownItem.propTypes = {
  uid: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
};

MultiSelectDropdownItem.defaultProps = {
  isSelected: false,
};
