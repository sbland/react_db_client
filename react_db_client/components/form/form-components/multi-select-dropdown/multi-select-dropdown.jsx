import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@samnbuk/react_db_client.components.emoji';

import './multiSelectDropdown.scss';


export const MultiSelectDropdownItem = ({
  uid,
  label,
  handleSelect,
  isSelected,
}) => (
  <li className="multiSelectDropdown_item">
    <button
      type="button"
      className={(isSelected) ? 'button-two selected' : 'button-one notSelected'}
      onClick={() => handleSelect(uid)}
    >
      {label}
    </button>
  </li>
);

MultiSelectDropdownItem.propTypes = {
  uid: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  handleSelect: PropTypes.func.isRequired,
  isSelected: PropTypes.bool,
};

MultiSelectDropdownItem.defaultProps = {
  isSelected: false,
};

export const MultiSelectDropdown = ({
  activeSelection,
  updateActiveSelection,
  options = [],
}) => {
  const [menuOpen, setMenuOpen] = useState(false);
  const menuRef = useRef(null);

  function handleClickOutside(event) {
    if (menuRef.current && !menuRef.current.contains(event.target)) {
      setMenuOpen(false);
    }
  }

  useEffect(() => {
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  });

  const handleSelect = (uid) => {
    const indexInSelection = activeSelection.indexOf(uid);
    const copyOfActiveSelection = [...activeSelection];
    if (indexInSelection !== -1) {
      copyOfActiveSelection.splice(indexInSelection, 1);
    } else {
      copyOfActiveSelection.push(uid);
    }
    const sortedList = options
      .map((opt) => opt.uid)
      .filter((opt) => copyOfActiveSelection.indexOf(opt) !== -1);
    updateActiveSelection(sortedList);
  };

  const selectedItems = activeSelection && activeSelection.map((sel) => (options.filter((opt) => opt.uid === sel)[0] || { uid: sel, label: sel }));
  const unselectedItems = (options && options.length > 0)
    ? options.filter((opt) => (activeSelection.indexOf(opt.uid) === -1))
    : [{ uid: 'invalid', label: 'There are no available options' }];

  const sortByLabel = (a, b) => (a.label.toUpperCase() > b.label.toUpperCase());

  const mapSelectedItems = selectedItems.sort(sortByLabel).map((opt) => (
    <MultiSelectDropdownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
      isSelected
    />
  ));

  const mapUnselectedItems = unselectedItems.sort(sortByLabel).map((opt) => (
    <MultiSelectDropdownItem
      key={opt.uid}
      uid={opt.uid}
      label={opt.label}
      handleSelect={handleSelect}
    />
  ));

  return (
    <div className="multiSelectDropdown">
      <button
        className="filterBtn"
        type="button"
        onClick={() => setMenuOpen(!menuOpen)}
      >
        <Emoji emoj="ðŸ”½" label="DropDownMenu" />
      </button>
      <ul className="multiSelectDropdown_list-selected">
        {mapSelectedItems}
      </ul>
      {menuOpen && (
        <div className="multiSelectDropdown_menu" ref={menuRef}>
          <ul className="multiSelectDropdown_list-unselected">
            {mapUnselectedItems}
          </ul>
        </div>
      )}
    </div>
  );
};

MultiSelectDropdown.propTypes = {
  activeSelection: PropTypes.arrayOf(PropTypes.string),
  options: PropTypes.arrayOf(PropTypes.shape({
    uid: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
  })).isRequired,
  updateActiveSelection: PropTypes.func.isRequired,
};

MultiSelectDropdown.defaultProps = {
  activeSelection: [],
};

