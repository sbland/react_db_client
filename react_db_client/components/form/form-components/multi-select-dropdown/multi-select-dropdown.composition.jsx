import React, { useState } from 'react';
import { MultiSelectDropdown } from './multi-select-dropdown';

export const demoOptions = [
  {
    uid: 'a',
    label: 'A',
  },
  {
    uid: 'b',
    label: 'B',
  },
  {
    uid: 'c',
    label: 'C',
  },
  {
    uid: 'd',
    label: 'D',
  },
  {
    uid: 'e',
    label: 'E',
  },
];

export const BasicMultiSelectDropdown = () => (
  <MultiSelectDropdown
    activeSelection={[]}
    updateActiveSelection={(newVal) => {}}
    options={demoOptions}
    required
  />
);

export const MultiSelectDropdownInteractive = () => {
  const [activeSelection, setActiveSelection] = useState(['a', 'b', 'e']);
  return (
    <MultiSelectDropdown
      activeSelection={activeSelection}
      options={demoOptions}
      updateActiveSelection={setActiveSelection}
    />
  );
};
