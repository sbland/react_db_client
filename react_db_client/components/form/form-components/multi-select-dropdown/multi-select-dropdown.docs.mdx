---
description: 'A MultiSelectDropdown component.'
labels: ['ui']
---

import { MultiSelectDropdown } from './multi-select-dropdown';

## React Component for rendering text

A basic div that renders some text

### Component usage
```js
<MultiSelectDropdown text="hello from MultiSelectDropdown" />
```

### Using props to customize the text

Modify the text to see it change live:
```js live
<MultiSelectDropdown text="hello from MultiSelectDropdown" />
```
