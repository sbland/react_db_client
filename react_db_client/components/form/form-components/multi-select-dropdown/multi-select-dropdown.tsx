import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@react_db_client/components.emoji';

import { MultiSelectDropdownItem } from './multi-select-dropdown-item';
import { IOpt } from './lib';
import { MultiSelectDropdownStyles } from './styles';

export interface MultiSelectDropdown {
  activeSelection: (string | number)[];
  updateActiveSelection: (newItem: (string | number)[]) => void;
  options: IOpt[];
  selectButtonProps?: React.ComponentProps<'button'>; //React.HTMLProps<HTMLButtonElement>;
}

export const MultiSelectDropdown = ({
  activeSelection,
  updateActiveSelection,
  options = [],
  selectButtonProps = {},
}: MultiSelectDropdown) => {
  const [menuOpen, setMenuOpen] = useState(false);
  const menuRef = useRef<HTMLDivElement>(null);

  function handleClickOutside(event) {
    if (menuRef.current && !menuRef.current.contains(event.target)) {
      setMenuOpen(false);
    }
  }

  useEffect(() => {
    // Bind the event listener
    document.addEventListener('mousedown', handleClickOutside);
    return () => {
      // Unbind the event listener on clean up
      document.removeEventListener('mousedown', handleClickOutside);
    };
  });

  const handleSelect = (uid: string | number) => {
    const indexInSelection = activeSelection.indexOf(uid);
    const copyOfActiveSelection = [...activeSelection];
    if (indexInSelection !== -1) {
      copyOfActiveSelection.splice(indexInSelection, 1);
    } else {
      copyOfActiveSelection.push(uid);
    }
    const sortedList = options
      .map((opt) => opt.uid)
      .filter((opt) => copyOfActiveSelection.indexOf(opt) !== -1);
    updateActiveSelection(sortedList);
  };

  const selectedItems =
    activeSelection &&
    activeSelection.map(
      (sel) => options.filter((opt) => opt.uid === sel)[0] || { uid: sel, label: sel }
    );
  const unselectedItems =
    options && options.length > 0
      ? options.filter((opt) => activeSelection.indexOf(opt.uid) === -1)
      : [{ uid: 'invalid', label: 'There are no available options' }];

  const sortByLabel = (a: IOpt, b: IOpt) =>
    a.label.toUpperCase() >= b.label.toUpperCase() ? 1 : -1;

  const mapSelectedItems = selectedItems
    .sort(sortByLabel)
    .map((opt) => (
      <MultiSelectDropdownItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label as string}
        handleSelect={handleSelect}
        isSelected
      />
    ));

  const mapUnselectedItems = unselectedItems
    .sort(sortByLabel)
    .map((opt) => (
      <MultiSelectDropdownItem
        key={opt.uid}
        uid={opt.uid}
        label={opt.label}
        handleSelect={handleSelect}
      />
    ));

  return (
    <MultiSelectDropdownStyles className="multiSelectDropdown">
      <button
        className="filterBtn"
        type="button"
        onClick={() => setMenuOpen(!menuOpen)}
        {...selectButtonProps}
      >
        <Emoji emoj="ðŸ”½" label="DropDownMenu" />
      </button>
      <ul className="multiSelectDropdown_list-selected">{mapSelectedItems}</ul>
      {menuOpen && (
        <div className="multiSelectDropdown_menu" ref={menuRef}>
          <ul className="multiSelectDropdown_list-unselected">{mapUnselectedItems}</ul>
        </div>
      )}
    </MultiSelectDropdownStyles>
  );
};

MultiSelectDropdown.propTypes = {
  activeSelection: PropTypes.arrayOf(PropTypes.string),
  options: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  updateActiveSelection: PropTypes.func.isRequired,
};

MultiSelectDropdown.defaultProps = {
  activeSelection: [],
};
