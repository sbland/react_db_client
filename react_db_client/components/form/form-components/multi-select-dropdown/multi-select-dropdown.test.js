import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { MultiSelectDropdown, MultiSelectDropdownItem } from './multi-select-dropdown';
import * as compositions from './multi-select-dropdown.composition';

describe('MultiSelectDropdown', () => {
  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        mount(<Composition />);
      });
    });
  });

  it('Renders', () => {
    shallow(
      <MultiSelectDropdown
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={() => {}}
      />
    );
  });
  it('Matches Snapshot', () => {
    const component = shallow(
      <MultiSelectDropdown
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={() => {}}
      />
    );
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('selecting', () => {
    const updateActiveSelection = jest.fn();
    const bubbleSelector = mount(
      <MultiSelectDropdown
        activeSelection={['a', 'b']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={updateActiveSelection}
      />
    );
    it('highlights selected', () => {
      expect(bubbleSelector.find('.selected').length).toEqual(2);
    });
    // TODO: Below does not work as unselected are hidden!
    // it('calls update selection with item added when non-selected item is clicked', () => {
    //   console.log(bubbleSelector.debug());
    //   const selectedItemBtn = bubbleSelector.find('.notSelected').first();
    //   selectedItemBtn.simulate('click');
    //   expect(updateActiveSelection).toHaveBeenCalled();
    //   expect(updateActiveSelection).toHaveBeenCalledWith(['a', 'b', 'c']);
    //   updateActiveSelection.mockClear();
    // });
    it('calls update selection with item removed when selected item is clicked', () => {
      const selectedItemBtn = bubbleSelector.find('.selected').first();
      selectedItemBtn.simulate('click');
      expect(updateActiveSelection).toHaveBeenCalled();
      expect(updateActiveSelection).toHaveBeenCalledWith(['b']);
      updateActiveSelection.mockClear();
    });
  });
  describe('Adding custom entry', () => {
    const updateActiveSelection = jest.fn();
    const bubbleSelector = mount(
      <MultiSelectDropdown
        activeSelection={['a', 'b', 'e']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={updateActiveSelection}
      />
    );
    it('allows selections outside of option set', () => {
      expect(bubbleSelector.find('.selected').length).toEqual(3);
    });
    // it('adds a selection on entering into text field and pressing enter', () => {
    //   // const textField =
    // });
  });
  describe('sorts selection', () => {
    const updateActiveSelection = jest.fn();
    const multiSelectDropDown = mount(
      <MultiSelectDropdown
        activeSelection={['a', 'c']}
        options={[
          { uid: 'a', label: 'a' },
          { uid: 'b', label: 'b' },
          { uid: 'c', label: 'c' },
          { uid: 'd', label: 'd' },
        ]}
        updateActiveSelection={updateActiveSelection}
      />
    );
    it('allows selections outside of option set', () => {
      const dropDownBtn = multiSelectDropDown.find('.filterBtn');
      dropDownBtn.simulate('click');
      const multiSelectDropdown_menu = multiSelectDropDown.find('.multiSelectDropdown_menu');
      const item = multiSelectDropdown_menu.find(MultiSelectDropdownItem);
      const btn = item.at(0).find('button');
      btn.simulate('click');
      expect(updateActiveSelection).toHaveBeenCalledWith(['a', 'b', 'c']);
    });
  });
});
