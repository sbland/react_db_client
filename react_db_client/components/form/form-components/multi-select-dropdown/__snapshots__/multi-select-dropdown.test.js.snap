// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectDropdown Matches Snapshot 1`] = `
"<styled.div className=\\"multiSelectDropdown\\">
  <button className=\\"filterBtn\\" type=\\"button\\" onClick={[Function: onClick]}>
    <Emoji emoj=\\"ðŸ”½\\" label=\\"DropDownMenu\\" className=\\"\\" />
  </button>
  <ul className=\\"multiSelectDropdown_list-selected\\">
    <MultiSelectDropdownItem uid=\\"a\\" label=\\"a\\" handleSelect={[Function: handleSelect]} isSelected={true} />
    <MultiSelectDropdownItem uid=\\"b\\" label=\\"b\\" handleSelect={[Function: handleSelect]} isSelected={true} />
  </ul>
</styled.div>"
`;
